### PR TITLE
feature: skid marks

### PR DESCRIPTION
**Pull request by njm222**

### CODE REVIEW

**Comment by drcmda** [comment]
Great! I tried almost the same with planes before. I wanted the skid to form an actual trail so I made the plane longer, but then I couldn't figure out the rotation, it was looking everywhere but the wheels. Do you know how to fix that?

**Comment by njm222** [comment]
Yeah I'm stuck on the rotation bit at the moment. I'll check in what I have at the moment, but I'm having the same issue trying to match the rotation of the skid marks with the wheels. I'm trying with the `chassisBody` instead of the `wheels`

**Comment by drcmda** [comment]
maybe we don't need it, i tried this and it looks pretty good:

```jsx
import * as THREE from 'three'
import { useRef } from 'react'
import { useFrame } from '@react-three/fiber'
import { useStore } from '../utils/store'

const v = new THREE.Vector3()
const m = new THREE.Matrix4()
const o = new THREE.Object3D()
const q = new THREE.Quaternion()

export function Skid({ opacity = 0.8 }) {
  const { wheels } = useStore((state) => state.raycast)
  const skid = useRef()

  let index = 0
  let time = 0
  let intensity = 0

  function setItemAt(target, obj, i, intensity) {
    o.position.set(obj.position.x, obj.position.y - 0, obj.position.z)
    o.rotation.set(-Math.PI / 2, Math.random(), Math.random())
    o.scale.setScalar(1)
    o.updateMatrix()
    target.setMatrixAt(i, o.matrix)
    skid.current.instanceMatrix.needsUpdate = true
  }

  useFrame((state, delta) => {
    const store = useStore.getState()
    const { controls, sliding, speed } = useStore.getState()
    intensity = THREE.MathUtils.lerp(intensity, (controls.brake * speed) / 40, delta * 8)

    if (controls.brake) {
      time = state.clock.getElapsedTime()
      // Set new skid
      setItemAt(skid.current, wheels[2].current, index++, intensity)
      setItemAt(skid.current, wheels[3].current, index++, intensity)
      if (index === 200) index = 0
    }
  })

  return (
    <instancedMesh ref={skid} args={[null, null, 200]}>
      <planeGeometry args={[0.4, 0.4]} />
      <meshBasicMaterial color="black" transparent opacity={opacity} depthWrite={false} />
    </instancedMesh>
  )
}
```

**Comment by njm222** [comment]
Looks great for now! I was trying something similar but didn't get it to look as good

**Comment by drcmda** [comment]
yep i think its good enough. sorry, small mistake above, one of the math.random was wrong. only in the z vec

**Comment by njm222** [comment]
yup fixing now and also adding the skid marks for `sliding`

**Comment by drcmda** [comment]
and with this:

```jsx
    const { controls, sliding } = useStore.getState()
    if (controls.brake && sliding) {
```

the skid stays on the road even when you continue pressing break :-D

**Comment by drcmda** [comment]
made a smiley 

Screenshot 2021-06-03 at 09 51 26

**Comment by njm222** [comment]
I tried something a little different with the sliding... let me know what you think!
I think it would work better if we had a sliding value for each weel

**Comment by drcmda** [comment]
i think i like it better if its just the backwheels and only on brake. the raycast vehicle is configured like this:

```jsx
    for (let e = 2; e < 4; e++) api.applyEngineForce(speed < maxSpeed ? engineValue : 0, e)
    const steeringValue = left || right ? dynamicSteer * (left && !right ? 1 : -1) : 0
    for (let s = 0; s < 2; s++) api.setSteeringValue(steeringValue, s)
    for (let b = 2; b < 4; b++) api.setBrake(brake ? (forward ? maxBrake / 1.5 : maxBrake) : 0, b)
```

so engine only drives the back wheels
steering only front wheels
brake only back wheels

**Comment by njm222** [comment]
Ahhh okay my understanding of `sliding` was incorrect. I'll revert back to your version

**Comment by njm222** [comment]
resolves #2

File: src/effects/Skid.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  useFrame(() => {
+    const { controls, sliding } = useStore.getState()
+
+    if (controls.brake && sliding) {
+      // Set new skid
+      setItemAt(skid, wheels[2].current, index++)
+      setItemAt(skid, wheels[3].current, index++)
+      if (index === skidLength) index = 0
```
Comment (williaster) [inline src/effects/Skid.js:32]: just trying to understand this (v new to three 😄 ): this means that the `Skid` consists of a max of 200 instances of a plane and you constantly move the trailing instances to the newest brake position if you've reached 200?

**Review by njm222 — COMMENTED** [review]
@williaster Yes pretty much what you said. You can think of `useFrame` as the render loop (the more computation you put in your render loop the slower / longer it will run).

So during each iteration,
when we want to draw a skid mark 
```jsx 
if (controls.brake && sliding)
```
We draw a plane under each of the back tires and increment `index`
```jsx
setItemAt(skid, wheels[2].current, index++)
setItemAt(skid, wheels[3].current, index++)
```

**Pull request merged**